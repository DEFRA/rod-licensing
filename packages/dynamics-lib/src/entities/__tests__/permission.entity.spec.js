import { Permission, Contact, Permit, PoclFile, Transaction, retrieveGlobalOptionSets } from '../../index.js'

let optionSetData
describe('permission entity', () => {
  beforeAll(async () => {
    optionSetData = await retrieveGlobalOptionSets().cached()
  })

  it('maps from dynamics', async () => {
    const permission = Permission.fromResponse(
      {
        '@odata.etag': 'W/"186695153"',
        defra_permissionid: '347a9083-361e-ea11-a810-000d3a25c5d6',
        defra_name: '00000000-2WC3FDR-CD379B',
        defra_issuedate: '2019-12-13T09:00:00Z',
        defra_startdate: '2019-12-14T00:00:00Z',
        defra_enddate: '2020-12-13T23:59:59Z',
        defra_stagingid: '71ad9a25-2a03-406b-a0e3-f4ff37799374',
        defra_datasource: 910400003,
        defra_renewal: true,
        defra_licenceforyou: 1
      },
      optionSetData
    )

    const expectedFields = {
      id: '347a9083-361e-ea11-a810-000d3a25c5d6',
      referenceNumber: '00000000-2WC3FDR-CD379B',
      issueDate: '2019-12-13T09:00:00Z',
      startDate: '2019-12-14T00:00:00Z',
      endDate: '2020-12-13T23:59:59Z',
      stagingId: '71ad9a25-2a03-406b-a0e3-f4ff37799374',
      dataSource: expect.objectContaining({ id: 910400003, label: 'Web Sales', description: 'Web Sales' }),
      isRenewal: true,
      isLicenceForYou: expect.objectContaining({ id: 1, label: 'Yes', description: 'Yes' })
    }

    expect(permission).toBeInstanceOf(Permission)
    expect(permission).toMatchObject(expect.objectContaining({ etag: 'W/"186695153"', ...expectedFields }))
    expect(permission.toJSON()).toMatchObject(expect.objectContaining(expectedFields))
    expect(JSON.parse(permission.toString())).toMatchObject(expect.objectContaining(expectedFields))
  })

  it('maps to dynamics', async () => {
    // Mimic a permit which is already persisted to test binding
    const permit = Permit.fromResponse(
      {
        '@odata.etag': 'W/"22639016"',
        defra_availablefrom: '2017-03-31T23:00:00Z',
        defra_availableto: '2020-03-31T22:59:00Z',
        defra_duration: 910400000,
        defra_permittype: 910400000,
        defra_advertisedprice: 6.0,
        defra_permitid: '9d1b34a0-0c66-e611-80dc-c4346bad0190',
        defra_name: '2017-20 Coarse 1 day 2 Rod Licence (Full)',
        defra_permitsubtype: 910400000,
        defra_equipment: 910400000,
        defra_isforfulfilment: false,
        defra_iscountersales: true,
        defra_advertisedprice_base: 6.0,
        defra_itemid: '42289'
      },
      optionSetData
    )
    // Mimic a new contact to test binding
    const contact = new Contact()
    // Mimic a new transaction to test binding
    const transaction = new Transaction()
    const poclFile = new PoclFile()

    const permission = new Permission()
    permission.referenceNumber = '00000000-2WC3FDR-CD379B'
    permission.issueDate = '2019-12-13T09:00:00Z'
    permission.startDate = '2019-12-14T00:00:00Z'
    permission.endDate = '2020-12-13T23:59:59Z'
    permission.stagingId = '71ad9a25-2a03-406b-a0e3-f4ff37799374'
    permission.dataSource = optionSetData.defra_datasource.options['910400003']
    permission.isRenewal = true
    permission.isLicenceForYou = optionSetData.defra_islicenceforyou.options['1']

    permission.bindToEntity(Permission.definition.relationships.licensee, contact)
    permission.bindToEntity(Permission.definition.relationships.permit, permit)
    permission.bindToEntity(Permission.definition.relationships.transaction, transaction)
    permission.bindToEntity(Permission.definition.relationships.poclFile, poclFile)

    const dynamicsEntity = permission.toRequestBody()
    expect(dynamicsEntity).toMatchObject(
      expect.objectContaining({
        defra_name: '00000000-2WC3FDR-CD379B',
        defra_issuedate: '2019-12-13T09:00:00Z',
        defra_startdate: '2019-12-14T00:00:00Z',
        defra_enddate: '2020-12-13T23:59:59Z',
        defra_stagingid: '71ad9a25-2a03-406b-a0e3-f4ff37799374',
        defra_datasource: 910400003,
        defra_licenceforyou: 1,
        'defra_PermitId@odata.bind': `/${Permit.definition.dynamicsCollection}(${permit.id})`,
        'defra_ContactId@odata.bind': `$${contact.uniqueContentId}`,
        'defra_Transaction@odata.bind': `$${transaction.uniqueContentId}`,
        'defra_POCLFileId@odata.bind': `$${poclFile.uniqueContentId}`,
        defra_renewal: true
      })
    )
  })
})
