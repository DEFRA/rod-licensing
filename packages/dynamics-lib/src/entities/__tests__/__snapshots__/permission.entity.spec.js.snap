// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`permission entity maps from dynamics matches JSON representation 1`] = `
Object {
  "dataSource": Object {
    "description": "Door to door sales",
    "id": "99999999",
    "label": "Door to door sales",
  },
  "endDate": "2020-12-13T23:59:59Z",
  "id": "347a9083-361e-ea11-a810-000d3a25c5d6",
  "isLicenceForYou": Object {
    "description": "Definitely maybe",
    "id": "1",
    "label": "Maybe",
  },
  "isMultiBuy": true,
  "issueDate": "2019-12-13T09:00:00Z",
  "referenceNumber": "00000000-2WC3FDR-CD379B",
  "stagingId": "71ad9a25-2a03-406b-a0e3-f4ff37799374",
  "startDate": "2019-12-14T00:00:00Z",
}
`;

exports[`permission entity maps from dynamics matches JSON representation 2`] = `
Object {
  "dataSource": Object {
    "description": "Door to door sales",
    "id": "99999999",
    "label": "Door to door sales",
  },
  "endDate": "2020-12-13T23:59:59Z",
  "id": "347a9083-361e-ea11-a810-000d3a25c5d6",
  "isLicenceForYou": Object {
    "description": "Definitely maybe",
    "id": "1",
    "label": "Maybe",
  },
  "isMultiBuy": true,
  "issueDate": "2019-12-13T09:00:00Z",
  "referenceNumber": "00000000-2WC3FDR-CD379B",
  "stagingId": "71ad9a25-2a03-406b-a0e3-f4ff37799374",
  "startDate": "2019-12-14T00:00:00Z",
}
`;

exports[`permission entity maps to dynamics 1`] = `
Object {
  "defra_ContactId@odata.bind": "$$45812ea6-991f-41b3-8760-03542f9cc133",
  "defra_POCLFileId@odata.bind": "$$45812ea6-991f-41b3-8760-03542f9cc131",
  "defra_PermitId@odata.bind": "/defra_permits(9d1b34a0-0c66-e611-80dc-c4346bad0190)",
  "defra_Transaction@odata.bind": "$$45812ea6-991f-41b3-8760-03542f9cc132",
  "defra_datasource": "99999999",
  "defra_enddate": "2020-12-13T23:59:59Z",
  "defra_ismultibuy": true,
  "defra_issuedate": "2019-12-13T09:00:00Z",
  "defra_licenceforyou": "1",
  "defra_name": "00000000-2WC3FDR-CD379B",
  "defra_renewal": true,
  "defra_stagingid": "71ad9a25-2a03-406b-a0e3-f4ff37799374",
  "defra_startdate": "2019-12-14T00:00:00Z",
}
`;
